src/Microsoft.CodeAnalysis.Razor.Workspaces/DefaultTagHelperResolver.cs

```diff
@@ -20,68 +19,42 @@ public override TagHelperResolutionResult GetTagHelpers(Compilation compilation,
         {
             var descriptors = new List<TagHelperDescriptor>();
 
-            VisitTagHelpers(compilation, assemblyNameFilters, descriptors);
-            VisitViewComponents(compilation, assemblyNameFilters, descriptors);
-
-            var diagnostics = new List<RazorDiagnostic>();
-            var resolutionResult = new TagHelperResolutionResult(descriptors, diagnostics);
-
-            return resolutionResult;
-        }
-
-        private void VisitTagHelpers(Compilation compilation, IEnumerable<string> assemblyNameFilters, List<TagHelperDescriptor> results)
-        {
-            var types = new List<INamedTypeSymbol>();
-            var visitor = TagHelperTypeVisitor.Create(compilation, types);
-
-            VisitCompilation(visitor, compilation);
+            var providers = new ITagHelperDescriptorProvider[]
+            {
+                new DefaultTagHelperDescriptorProvider(),
+                new ViewComponentTagHelperDescriptorProvider(),
+            };
 
-            var factory = new DefaultTagHelperDescriptorFactory(compilation, DesignTime);
+            var results = new List<TagHelperDescriptor>();
+            var context = new Context(results);
 
-            foreach (var type in types)
+            for (var i = 0; i < providers.Length; i++)
             {
-                if (assemblyNameFilters == null || assemblyNameFilters.Contains(type.ContainingAssembly.Identity.Name))
-                {
-                    var descriptor = factory.CreateDescriptor(type);
-
-                    if (descriptor != null)
-                    {
-                        results.Add(descriptor);
-                    }
-                }
+                var provider = providers[i];
+                provider.Execute(context);
             }
+            
+            var diagnostics = new List<RazorDiagnostic>();
+            var resolutionResult = new TagHelperResolutionResult(results, diagnostics);
+
+            return resolutionResult;
         }
 
-        private void VisitViewComponents(Compilation compilation, IEnumerable<string> assemblyNameFilters, List<TagHelperDescriptor> results)
+        private class Context : TagHelperDescriptorProviderContext
```

Comment:
Even though this bit is purely for tooling it might make sense to have a `TagHelperDescriptorProvierContext.CreateDefault` since we have a duplication in the TagHelperFeature. I prefer that approach because it limits potential bugs across runtime/tooling experiences.